### PR TITLE
Fixes bugs

### PR DESCRIPTION
**Pull request by ashishnegi**
Depends on PR #7 
1. Flushes incoming channel before starting the game; Moves Chan to TChan  
2. Sends clientId to the UI so that it can know its state / for debugging.
3. Removes playClient thread duplicating msgs to websockets as now websockets have outChan itself.
4. Client view now has (0,0) at top-left.
5. Player head is in green color, red when he dies. 
6. Starts another game after one finishes.
7. Makes client sesion sticky; Can replay after game finishes.
8. Player starts at random position and orientation.
9. Adds tcp sockets to the server. Now supports both websockets and tcp.

### CODE REVIEW

File: src/Tronkell/Server/Server.hs [hunk 1]
```diff
@@ -38,12 +39,14 @@ startServer gConfig = do
     moveAllPlayersToWaiting (serverUsers server)
     gameEngineThread server Nothing
 
-  -- start websockets
-  wsThread <- forkIO $ W.start firstUId networkChans clientsChan
+  -- start network
+  wsThread  <- forkIO $ W.start firstUId networkChans clientsChan
+  tcpThread <- forkIO $ Tcp.start firstUId networkChans clientsChan
```
Comment (ford-prefect) [inline src/Tronkell/Server/Server.hs]: I guess this is the thing that can become our "interface" for transports at some point.

File: src/Tronkell/Types.hs [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
+        bInt = (fromEnum b)
+        diffAB = bInt - aInt + 1
+        n' = (n `rem` diffAB) + aInt
+        orien = case n' of
+                  0 -> North
+                  1 -> East
+                  2 -> South
+                  3 -> West
+                  _ -> North
+    in (orien, gen')
```
Comment (ford-prefect) [inline src/Tronkell/Types.hs]: Don't change this now, but this looks like a stray code that should have been in the previous commit made its way in.

File: src/Tronkell/Server/Server.hs [hunk 3]
```diff
@@ -42,8 +42,13 @@ startServer gConfig = do
   -- start network
   wsThread  <- forkIO $ W.start firstUId networkChans clientsChan
   tcpThread <- forkIO $ Tcp.start firstUId networkChans clientsChan
+  -- to avoid space msg leak..
+  forkIO . forever . readChan $ clientSpecificOutChan
+  forkIO . forever . readChan $ clientsChan
+  forkIO . forever . readChan $ internalChan
```
Comment (ford-prefect) [inline src/Tronkell/Server/Server.hs]: Could you expand on why this is needed?

File: src/Tronkell/Types.hs (hunk 2 above)
Comment (ford-prefect) [inline src/Tronkell/Types.hs]: That sounds awkward. Isn't there a better broadcast mechanism? @abhin4v ?

File: src/Tronkell/Server/Server.hs (hunk 3 above)
Comment (ford-prefect) [inline src/Tronkell/Server/Server.hs]: I think github ate my reply, the last time. This feels a bit clunky. Do we not have a better way to do broadcast?

Either way, I think this is okay to push and add as a FIXME. Don't want to block the other changes you have.

Comment (ford-prefect) [inline src/Tronkell/Server/Server.hs]: @abhin4v ^ comments on broadcast mechanism, please.

File: src/Tronkell/Types.hs (hunk 2 above)
Comment (ford-prefect) [inline src/Tronkell/Types.hs]: Ugh, my bad, sorry.

File: src/Tronkell/Server/Server.hs (hunk 3 above)
Comment (ashishnegi) [inline src/Tronkell/Server/Server.hs]: there is something like : https://hackage.haskell.org/package/stm-2.4.4.1/docs/Control-Concurrent-STM-TChan.html#v:newBroadcastTChan
and 
there is something like https://hackage.haskell.org/package/broadcast-chan
quoting the library :
```
A variation of Control.Concurrent.Chan from base, which allows to the easy creation of broadcast channels without the space-leaks that may arise from using Control.Concurrent.Chan.dupChan. The Control.Concurrent.Chan.Chan type from Control.Concurrent.Chan consists of both a read and write end. This presents a problem when one wants to have a broadcast channel that, at times, has zero listeners. To write to a Control.Concurrent.Chan.Chan there must always be a read end and this read end will hold ALL messages alive until read. The simple solution applied in this module is to separate read and write ends. As a result, any messages written to the write end can be immediately garbage collected if there are no active read ends, avoding space leaks.
```

but they both required some more work then these 3 lines.

Comment (ford-prefect) [inline src/Tronkell/Server/Server.hs]: (but again, I think this can be a FIXME and really doesn't need to block this)

Comment (ashishnegi) [inline src/Tronkell/Server/Server.hs]: i tried creating a `broadcast fn`.. but writes are to `duplicate chan` and at that point we do/may not have access to main channel.. 
i am adding a fixme for now.

**Pull request merged**